lib/hotel.rb

```diff
@@ -0,0 +1,183 @@
+require_relative "reservation"
+require_relative "date_range"
+require_relative "hotel_block"
+require "date"
+
+module HotelSystem
+  class Hotel
+    attr_reader :rooms, :blocks
+    attr_accessor :reservations
+
+    def initialize(number_of_rooms)
+      @rooms = []
+      add_rooms(number_of_rooms)
+      @reservations = []
+      @blocks = []
+    end
+
+    def add_rooms(number)
+      if number.class != Integer || number < 1
+        raise ArgumentError, "Please enter a number greater than 1."
+      end
+      (1..number).each do |num|
+        room = HotelSystem::Room.new(num)
+        @rooms << room
+      end
+    end
+
+    def valid_date_entry?(year, month, day)
+      HotelSystem::DateRange.valid_date_entry?(year, month, day)
+    end
+
+    def create_date_range(start_year:, start_month:, start_day:, num_nights: nil)
+      valid_date_entry?(start_year, start_month, start_day)
+      dates = HotelSystem::DateRange.new(start_year: start_year, start_month: start_month, start_day: start_day, num_nights: num_nights)
+      return dates
+    end
+
+    def find_room(number)
+      HotelSystem::Room.valid_room_number(number)
+      return @rooms.find { |room| room.room_number == number }
+    end
+
+    def room_reserved?(room_number:, year:, month:, day:)
+      valid_date_entry?(year, month, day)
+      res_room = find_room(room_number)
+      res_date = Date.new(year, month, day)
+      if res_room.reservations == []
+        return false
+      end
+      res_room.reservations.each do |res|
+        if res.date_range.overlap?(res_date)
+          return true
+        end
+      end
+      return false
+    end 
+
+    def find_available_room(start_year:, start_month:, start_day:)
+      valid_date_entry?(start_year, start_month, start_day)
+      @rooms.each do |hotel_room|
+        reserved = room_reserved?(room_number: hotel_room.room_number, year: start_year, month: start_month, day: start_day)
+        block_status = hotel_room.in_block?(year: start_year, month: start_month, day: start_day)
+        if !reserved && !block_status
+          return hotel_room
+        end
+      end
+      raise NotImplementedError, "Sorry, no available rooms for that date."
+    end
+
+    def create_reservation_id
+      if @reservations.length == nil
+        return 1
+      else
+        id = @reservations.length + 1
+        return id
+      end
+    end
+
+    def reserve_room(start_year:, start_month:, start_day:, num_nights:)
+      valid_date_entry?(start_year, start_month, start_day)
+      res_dates = create_date_range(start_year: start_year, start_month: start_month, start_day: start_day, num_nights: num_nights)
+      # res_dates_range = res_dates.date_list
+      res_room = find_available_room(start_year: start_year, start_month: start_month, start_day: start_day)
+      id = create_reservation_id
+      res = HotelSystem::Reservation.new(id: id, room: res_room, date_range: res_dates)
+      @reservations << res
```

Comment:
Since `find_available_room` only looks at one day, it's possible that this code will double-book a room if the existing reservation starts after the first day of the new reservation.